### ISSUE TITLE
MPEG, MPEG-TS, MPEG-PS Support

### ISSUE THREAD

**Issue by AeroBuffalo** [jcodec/jcodec#181]
I am trying to be able to pick out frames (video and metadata) from MPEG, MPEG-TS and MPEG-PS files and streams (network / UDP / RTP streams). I was looking into using JCODEC to do this and I started off by trying to use the FrameGrab / FrameGrab8Bit classes, and ran into an error that those files are "temporarily unsupported". I looked into going back some commits to see if I could just use older code, but it looks like both of those files have had those formats "temporarily unsupported" since 2103 / 2015, respectively. 

Are any of those formats supported somewhere in the library and I am just using the wrong interface? Or do I have to go and hook things together to be able to get video and metadata out of the files & streams? 

I saw issue #27 and I was not sure if that is related. If it is, is there any update on the progress? For example files, I am testing with sample files available from FFMPEG (http://samples.ffmpeg.org/MPEG2/mpegts-klv/).

I have started trying to plug things back in, in the FrameGrab8Bit class by putting in the following code, but I run into a packet header assertion failure in the MTSDemuxer.java file (information below code)

---FrameGrab8Bit additional code---
public static FrameGrab8Bit createFrameGrab8Bit(SeekableByteChannel _in) throws IOException, JCodecException {
...
        case MPEG_PS:
        	MPSDemuxer psd = new MPSDemuxer(_in);
        	List<MPEGDemuxerTrack> tracks = psd.getVideoTracks();
        	videoTrack_ = (SeekableDemuxerTrack)tracks.get(0);
//            throw new UnsupportedFormatException("MPEG PS is temporarily unsupported.");
        	break;
        case MPEG_TS:
        	MTSDemuxer tsd = new MTSDemuxer(_in);
        	MPSDemuxer ptsd = new MPSDemuxer(tsd.getProgram(0));
        	List<MPEGDemuxerTrack> tstracks = ptsd.getVideoTracks();
        	videoTrack_ = (SeekableDemuxerTrack)tstracks.get(0);
//            throw new UnsupportedFormatException("MPEG TS is temporarily unsupported.");
        	break;
...
---End FrameGrab8Bit additional code---

---MTSDemuxer.java Assertion failure---
 public static MTSPacket parsePacket(ByteBuffer buffer) {
    	int marker = buffer.get() & 0xff;
    	int marker = by & 0xff;
        Assert.assertEquals(0x47, marker);
...
---End MTSDemuxer.java Assertion failure---

**Comment by AeroBuffalo** [comment]
Also, where would one go to get the metadata out of those files? Is there something that would be able to get me both the metadata AND the image frame all in one? Or do I have to use two separate classes?

**Comment by AeroBuffalo** [comment]
Has anyone taken a look at this? @svitvitskiy, is this the same problem discussed in #27?

**Comment by zhuker** [comment]
you are very welcome to submit a pull request for the feature

**Comment by AeroBuffalo** [comment]
@zhuker When you say to submit a pull request, do you mean just a blank request for the feature? Or to implement the feature and then submit the request? 

If you mean the former, do you know when about when that feature would be implemented? And if you mean the later, would you know a rough outline of what all would need to be done? I have been going through the code, and it seems like (at a minimum) the MTSDemuxer class would need to be recoded to be built off of the SeekableDemuxerTrack class. Is this correct?

And just to confirm, in either case, this does mean that the MPEG-TS demuxing is NOT currently fully implemented in the code?

Thanks,
Aero

**Comment by zhuker** [comment]
Yeah, MPEG TS is not supported to the extent MP4 is.
You can't really seek in TS streams (unless you index the entire stream
before hand). So it doesn't look like there's much sense to implement TS
demuxer on top of SeekableDemuxerTrack.
We haven't given much attention to TS demuxer actually, so any input is
very welcome.
Re PR - jcodec is a non profit project done on spare time of mostly
@svitvitskiy and less so myself, so please submit a PR with code :)

Cheers

**Comment by svitvitskiy** [comment]
@AeroBuffalo, yes MPEG TS is supported but not in the FrameGrab line of interfaces. You can use MPEGTS with [Transcoder](https://github.com/jcodec/jcodec/blob/master/src/main/java/org/jcodec/api/transcode/Transcoder.java) and [TranscoderMain](https://github.com/jcodec/jcodec/blob/master/src/main/java/org/jcodec/api/transcode/TranscodeMain.java). These classes implement a generic transcoder between the input and output formats. Ideally I would like to then implement the FrameGrab and SequenceEncoder family of interfaces through Transcoder internally. This can be done by using a specialized source or sink in the transcoder (I am not even sure if source and sink abstractions have made it into Transcoder yet :) ).

@AeroBuffalo, MPEGTS can not implement the Demuxer interface directly since it is a container for a bunch of streams (called programs). Yes, in real life you usually get MPEG TS with what you might think is one stream (through it might not be the case when you dive into details) however in broadcast one TS stream can carry up to 27 Mbit/s worth of data (which can be multiple independent TV channels depending on the actual compression and video resolution). So while I agree that for the average user the MTS interface has to be simplified I can't quite imagine how because the complexity of the interface is caused by the complexity of the format itself. One idea I have is to just create a facade demuxer called MTSSimpleDemuxer that will just assume a file with one program and will try to figure out it's streams as best as it can. Let me know if you think this is a good idea.

**Comment by AeroBuffalo** [comment]
@svitvitskiy I have been able to use MPEGTS-Streamer to parse out the TS bits. Once I strip that away, I get down to the PES packet (within the TS payload). Can I then plug this into JCODEC directly and have it parse out the contents of the PES packet? If so, how can I go about doing this? Do I need to combine everything into a single frame and then use JCODEC to decode?

**Comment by B-R-Bender** [comment]
@AeroBuffalo did you manage to solve this in any way?

**Comment by AeroBuffalo** [comment]
@B-R-Bender Sorry, it has been so long I do not remember what I was able to do / get done.
